pandas/core/groupby/numba_.py

```diff
@@ -0,0 +1,138 @@
+import inspect
+import types
+
+import numpy as np
+
+from pandas.compat._optional import import_optional_dependency
+
+
+class InvalidApply(Exception):
+    pass
+
+
+def execute_groupby_function(splitter, f):
+    """Mimics apply_frame_axis0 which is the Cython equivalent of this function."""
+    results = []
+    for _, group in splitter:
+        # TODO: what about series names/dataframe columns
+        index = group.index
+        values_as_array = group.to_numpy()
+        index_as_array = index.to_numpy()
+        try:
+            # TODO: support *args, **kwargs here
+            group_result = f(values_as_array, index_as_array)
+        except Exception:
+            # We can't be more specific without knowing something about `f`
+            # Like we do in Cython
+            raise InvalidApply("Let this error raise above us")
+        # Reconstruct the pandas object (expected downstream)
+        # This construction will fail is there is mutation,
+        # but we're banning it with numba?
+        group_result = group._constructor(group_result, index=index)
+        results.append(group_result)
+
+    return results
+
+
+def validate_apply_function_signature(func):
+    """
+    Validate that the apply function's first 2 arguments are 'values' and 'index'.
+
+    func : function
+        function to be applied to each group and will be JITed
+    """
+    apply_function_signature = list(inspect.signature(func).parameters.keys())[:2]
+    if apply_function_signature != ["values", "index"]:
+        raise ValueError(
+            "The apply function's first 2 arguments must be 'values' and 'index'"
+        )
+
+
+def make_groupby_apply(
```

Comment:
can we share pandas/core/window/numba_.py at all? maybe put common functions in pandas/core/algos/numba_.py (we have another issue about creating this and moving pandas/core/aggregation.py and algorithms.py to there

Reply:
Definitely. Will work on that in another commit.

Reply:
Definitely, will consolidate in another commit.